consensus/istanbul/backend/engine.go

```diff
@@ -373,12 +372,10 @@ func (sb *backend) Finalize(chain consensus.ChainReader, header *types.Header, s
 			logger.Trace(logMsg, "header.Number", header.Number.Uint64(), "node address", sb.address, "rewardbase", header.Rewardbase)
 		}
 
-		if proposer.Weight() != 0 {
```

Comment:
This code checks if addressBook is activated or not.
It calls proposer.Weight() because Weight is 0 when addressBook is not activated yet.
but it is redundant as we can know it by checking data of stakingInfo.

Reply:
I'm not saying that you need to modify this PR, however, this change does not look related to this deep copy change. Please separate unrelated changes in your future works.

Reply:
WeightedCouncil has list of validators. and proposer is one of validators.
now every block has their own validator list. so the proposer object is changed for every block.
while I edit the code, I encountered a problem with this logic so I removed it. but the problem is solved by line 549 on weighted.go

Reply:
Thanks for your explanation =) !!